자판기 앱 5단계 - 관찰자(Observer) 패턴

popsmile opened the pull request:
- **음료추가** 와 **잔액추가** 액션이 일어날 때, `VendingMachine` 이 해당하는 `Notification` 을 `NotificationCenter` 에  post하도록 추가했습니다.
- `ViewController` 가 위의 `Notification` 을 지켜볼 수 있도록 `addObserver()` 해주었습니다. `NotificationCenter` 에서 특정 `Notification` 을 보내주면 실행할 `selector` 또한 지정해주었습니다.
- 이미지 뷰 모서리를 둥그렇게 하는 메소드를 `ViewController` 에서 삭제하고, `RoundedCornersImageView` 커스텀 뷰를 생성하여 생성초기화될 때 수행되도록 수정했습니다.
- 학습하고 추가한 내용을 README에 정리했습니다.

popsmile committed:
Add postNotification to VendingMachine

The VendingMachine posts two notification of which names are didAddBeverage and didInsertMoney to the NotificaionCenter

popsmile committed:
Remove the unnecessary method

popsmile committed:
Delete the re-declaration of the NotificationCenter instance

popsmile committed:
Add the ViewController as an observer of the notification didAddBeverage

popsmile committed:
Add the ViewController as an observer of the notification didInsertMoney

popsmile committed:
Arrange the order of the methods in ViewController

popsmile committed:
Create a custom ImageView with rounded corners

Separate the View's logic and the ViewController's one to use the MVC pattern.

popsmile committed:
Fix the indents of addObserver methods

popsmile committed:
Update README with the descriptions of the observer pattern

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -19,49 +19,43 @@ class ViewController: UIViewController {
+        registerAsObserver()
         showQuantities()
-        vendingMachine?.showBalance(with: balanceForm)
+        showBalance()
```
옵저버를 등록하고 노티 받아서 처리하는 showQuantities() 나 showBalance()를 직접 호출하는데
초기값을 표시하기 위해 직접 호출하는 방법 외에 다른 방법은 뭐가 있을까요?

popsmile replied:
피드백 주신 내용에 대해서 여러 방향으로 고민하고 시도해본 후에 구현한 내용입니다~!
- `vendingMachineWillAppear` 라는 `Notification` 을 추가했습니다.
- 자판기 객체 안에 `willAppear()` 메소드를 추가하고, 뷰 컨트롤러 `viewDidLoad()` 메소드에서 호출합니다.
- `registerAsObserver()` 에 뷰 컨트롤러를 해당 노티피케이션 관찰자로 추가해주고, 뷰가 로드된 시점에 딱 한 번만 일어나는 노티피케이션이므로 `block` 클로저에서 `removeObserver()` 로 관찰자 역할을 해제해주었습니다. 
  
아래는 고민하는 과정에서 생각했던 방향입니다. 위처럼 최종 결정한 방향과 아래 생각한 다른 방안들 (아니면, 다른 무언가) 중에 어떤 것이 더 알맞은지 확신이 안들어 적어보았습니다..!
- `VendingMachine.shared` 변수 인스턴스를 처음 접근시에 `postNotification` 을 수행하도록 추가했었습니다. 하지만 뷰컨트롤러가 관찰자로 추가되는 `viewDidLoad()` 메소드 호출 전에, `init()` 에서 자판기 인스턴스를 먼저 접근하기때문에 관찰자로 등록되기 전에 노티피케이션이 포스트되는 상황이 되어버렸습니다...
- `ViewController.viewDidLoad()` 메소드 내에서 직접 `postNotification` 을 수행하도록 추가했었습니다. 하지만, 객체 자신이 노티피케이션을 보내고 자신이 받아서 처리하는 것은 의미없는 이상한(?) 구조인 듯 하여 삭제했습니다...

popsmile committed:
Add a Notification named VendingMachineWillAppear

When the VendingMachine call the method named willAppear(), it posts the notification. The method is called in viewDidLoad() in the ViewController. After receiving the notification, the labels related the VendingMachine will be updated as the present data.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -22,15 +22,17 @@ class ViewController: UIViewController {
         registerAsObserver()
-        showQuantities()
-        showBalance()
+        vendingMachine?.willAppear()
```
시도는 좋았네요. 사실 vendingMachien.willApear() 에서 post만 대신할 뿐 viewDidLoad에서 처리하는 것과 다름이 없네요. 

모델 객체 생성, 옵저버 등록, 노티 전송, 핸들러 처리 흐름이 시간 순서대로 될 수 있는지 고민하는건 늘 반복해야 할 겁니다. 
ViewController 에서 할 일과 Model 에서 할 일을 잘 구분해보세요.
